test/rekt/features/knconf/event_patterns.go

```diff
@@ -0,0 +1,152 @@
+	"github.com/google/go-cmp/cmp"
+	v1 "knative.dev/eventing/pkg/apis/duck/v1"
+	"knative.dev/reconciler-test/pkg/eventshub"
+	"knative.dev/reconciler-test/pkg/feature"
+)
+
+type EventPattern struct {
+	Success  []bool
+	Interval []uint
+}
+
+func ReceivedEventsFromProber(ctx context.Context, prober *eventshub.EventProber, prefix string) []conformanceevent.Event {
+	events := prober.ReceivedOrRejectedBy(ctx, prefix)
+	sort.Slice(events, func(i, j int) bool {
+		return events[i].Time.Before(events[j].Time)
+	})
+	ret := make([]conformanceevent.Event, len(events))
+	for i, e := range events {
+		ret[i] = EventToEvent(e.Event)
+	}
+	return ret
+}
+
+func EventToEvent(in *cloudevents.Event) conformanceevent.Event {
+	if in == nil {
+		return conformanceevent.Event{}
+	}
+	out := conformanceevent.Event{
+		Attributes: conformanceevent.ContextAttributes{
+			SpecVersion:     in.SpecVersion(),
+			Type:            in.Type(),
+			ID:              in.ID(),
+			Source:          in.Source(),
+			Subject:         in.Subject(),
+			DataSchema:      in.DataSchema(),
+			DataContentType: in.DataContentType(),
+			Extensions:      make(conformanceevent.Extensions),
+		},
+		Data: string(in.Data()),
+	}
+	if !in.Time().IsZero() {
+		out.Attributes.Time = types.Timestamp{Time: in.Time()}.String()
+	}
+	for key, ext := range in.Extensions() {
+		if value, err := types.Format(ext); err == nil {
+			out.Attributes.Extensions[key] = value
+		}
+	}
+
+	return out
+}
+
+func PatternFromProber(ctx context.Context, prober *eventshub.EventProber, prefix string) EventPattern {
+	events := prober.ReceivedOrRejectedBy(ctx, prefix)
+	sort.Slice(events, func(i, j int) bool {
+		return events[i].Time.Before(events[j].Time)
+	})
+	got := EventPattern{
+		Success:  make([]bool, 0),
+		Interval: make([]uint, 0),
+	}
+	for i, event := range events {
+		got.Success = append(got.Success, event.Kind == eventshub.EventReceived)
+		if i == 0 {
+			got.Interval = []uint{0}
+		} else {
+			diff := events[i-1].Time.Unix() - event.Time.Unix()
+			got.Interval = append(got.Interval, uint(diff))
+		}
+	}
+	return got
+}
+
+func PatternFromEstimates(attempts, failures uint) EventPattern {
+	p := EventPattern{
+		Success:  []bool{},
+		Interval: []uint{},
+	}
+	for i := uint(0); i < attempts; i++ {
+		if i >= failures {
+			p.Success = append(p.Success, true)
+			p.Interval = append(p.Interval, 0) // TODO: calculate time.
+			return p
+		}
+		p.Success = append(p.Success, false)
+		p.Interval = append(p.Interval, 0) // TODO: calculate time.
+	}
+	return p
+}
+
+func AssertEventPatterns(prober *eventshub.EventProber, expected map[string]EventPattern) feature.StepFn {
+	return func(ctx context.Context, t feature.T) {
+		for prefix, want := range expected {
+			got := PatternFromProber(ctx, prober, prefix)
+
+			t.Logf("Expected Events %s; \n Got: %#v\nWant: %#v", prefix, got, want)
+
+			// Check event acceptance.
+			if len(want.Success) != 0 && len(got.Success) != 0 {
```

Comment:
this is a bug here:
https://github.com/knative/eventing/pull/5325

Reply:
fixed: https://github.com/knative/eventing/pull/5296/files#diff-918fd77970216177020a2666222af2f88ecc9a17f69d275e16e396a3e7e32d26R123